src/components/user-defined-sets/sets.js

```diff
@@ -0,0 +1,159 @@
+          ...state,
+          namedSets: state.namedSets.delete(key),
+        };
+      }
+      return state;
+    }
+
+    /**
+     * Deletes set entries for all named sets.
+     * @param {object} state The current state.
+     * @returns {object} The new state.
+     */
+    static deleteAllNamedSets(state) {
+      return {
+        ...state,
+        namedSets: ImmutableOrderedMap(),
+      };
+    }
+
+    /**
+     * Deletes the set mapping for the set specified using prevKey,
+     * then maps the set value to nextKey.
+     * @param {object} state The current state.
+     * @param {string} prevKey The previous key for the set.
+     * @param {string} nextKey The new key for the set.
+     * @returns {object} The new state.
+     */
+    static renameNamedSet(state, prevKey, nextKey) {
+      if (state.namedSets.has(prevKey)) {
+        const prevSet = state.namedSets.get(prevKey);
+        return {
+          ...state,
+          namedSets: state.namedSets.delete(prevSet).set(nextKey, prevSet),
+        };
+      }
+      return state;
+    }
+
+    /**
+     * Modifies the set value for the current set.
+     * @param {object} state The current state.
+     * @param {Set} set The new current set value.
+     * @returns {object} The new state.
+     */
+    static setCurrentSet(state, set) {
+      return {
+        ...state,
+        currentSet: ImmutableSet(set),
+      };
+    }
+
+    /**
+     * Sets the current set value to an empty set.
+     * @param {object} state The current state.
+     * @returns {object} The new state.
+     */
+    static clearCurrentSet(state) {
+      return {
+        ...state,
+        currentSet: ImmutableSet(),
+      };
+    }
+
+    /**
+     * Copies the current set to a named set entry.
+     * @param {object} state The current state.
+     * @param {string} key The key for the named set mapping.
+     * @param {boolean} clear Whether to clear the current set after the copying. Default: false
+     * @returns {object} The new state.
+     */
+    static nameCurrentSet(state, key, clear) {
+      return {
+        ...state,
+        namedSets: state.namedSets.set(key, state.currentSet),
+        currentSet: (clear ? ImmutableSet() : state.currentSet),
+      };
+    }
+
+    static serialize(state) {
+      const {
+        namedSets = ImmutableOrderedMap(),
+        currentSet = ImmutableSet(),
+      } = state;
+      return {
+        namedSets: namedSets.toArray(),
+        currentSet: currentSet.toArray(),
+      };
+    }
+
+    static unserialize(serializedState) {
+      const {
+        namedSets = [],
+        currentSet = [],
+      } = serializedState;
+      return {
+        namedSets: ImmutableOrderedMap(namedSets),
+        currentSet: ImmutableSet(currentSet),
+      };
+    }
+
+    static persist(state, setTypeKey, datasetKey) {
+      const serializedState = Sets.serialize(state);
+      const serializedSets = JSON.parse(localStorage.getItem('sets')) || {};
```

Comment:
This is getting a little ahead of things, but would we prefer to namespace this by dataset-id?